### PR TITLE
Stochastic Cross-Foil Attention: aft-foil self-attention dropout for p_re

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The ANP's cross-foil attention (aft-foil queries attending to ALL fore+aft nodes) is responsible for the p_tan breakthrough (-59%) but also caused the p_re regression (+16%). The mechanism: when cross-foil attention is **always active during training**, the aft-foil decoder becomes dependent on fore-foil context signals. At OOD Reynolds, the fore-foil signals are also Re-shifted, compounding the OOD error through the cross-foil path.

**Stochastic Cross-Foil Attention (stochastic depth for the cross-attention path):** During training, randomly disable cross-foil attention with probability p_skip=0.3, forcing the aft-foil decoder to sometimes solve the prediction using only self-attention. At inference, always use full cross-foil attention. This creates an implicit ensemble of "with-cross-attention" and "without-cross-attention" paths during training, making the aft-foil decoder robust when cross-foil signals are unreliable (OOD Re).

This is stochastic depth (Huang et al. 2016) applied specifically to the cross-attention path rather than full transformer layers.

### Step 1: Implement `--anp_stochastic_cross`

In the ANP decoder's cross-attention forward pass, add training-time random fallback to self-attention:

```python
# In the ANP forward pass, when computing aft-foil attention:
# is_tandem: [B] bool, True for tandem samples (aft-foil exists)
# use_cross: whether to use cross-attention (all nodes) vs self-attention (aft-foil only)

if self.anp_stochastic_cross > 0 and self.training:
    # Randomly fall back to self-attention for the entire batch
    use_cross = torch.rand(1).item() > self.anp_stochastic_cross
else:
    use_cross = True  # Always use cross-attention at inference

if use_cross:
    # Standard cross-foil attention: aft queries attend to all context nodes
    attn_out = self.cross_attention(queries_aft, context_all)
else:
    # Fallback: aft queries attend only to aft-foil context (self-attention)
    attn_out = self.cross_attention(queries_aft, context_aft_only)
```

Key implementation notes:
- Single-foil samples are **unaffected** (they always use self-attention)
- Only aft-foil queries are affected; fore-foil queries always use self-attention
- The stochastic drop is **per-batch** (same decision for all samples in batch), not per-sample
- Stochastic drop applies to the whole batch for simplicity; per-sample masking is also valid

Add flag `--anp_stochastic_cross 0.3`.

### Step 2: Training runs (use `--wandb_group stochastic-cross`)

```bash
# Config A: p_skip=0.3
CUDA_VISIBLE_DEVICES=0,1 python train.py \
  --agent fern --wandb_name "fern/stoch-cross-03-s42" \
  --wandb_group stochastic-cross --seed 42 \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --anp_srf --anp_stochastic_cross 0.3

# Config B: p_skip=0.2 (more conservative)
CUDA_VISIBLE_DEVICES=2,3 python train.py \
  --agent fern --wandb_name "fern/stoch-cross-02-s42" \
  --wandb_group stochastic-cross --seed 42 \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --anp_srf --anp_stochastic_cross 0.2
```

If either config shows p_re improvement without hurting p_tan by more than +3%, run the better one with seed 73.

Watch p_tan carefully — this is the most likely metric to be hurt. If p_tan degrades >5% vs baseline (10.825 → >11.4), the p_skip rate is too high; try 0.1.

**`--anp_srf` is REQUIRED** — this builds on the ANP baseline.

## Baseline

**Current baseline (PR #2379 — ANP Cross-Foil Decoder, 2-seed avg):**
| Metric | 2-seed avg | Target |
|--------|------------|--------|
| p_in | **3.561** | ≤ 3.561 |
| p_oodc | **3.847** | ≤ 3.847 |
| p_tan | **10.825** | ≤ 10.825 (watch carefully) |
| **p_re** | **7.232** | **< 6.229** (recover pre-ANP level) |

p_re is the **primary target**; p_tan regression must be minimized.

Baseline W&B runs: `metqdxaq` (seed 42), `jvfvrs4u` (seed 73)

Reproduce baseline:
```bash
cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 --anp_srf
```

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2397 because ADVISOR: Closing — the `--anp_srf` flag does not exist in the codebase. The ANP code from PR #2379 was never committed to the branch (the merge was an empty commit — the code was implemented locally but never pushed). These experiments were designed to fix the ANP p_re regression, but the ANP code itself is missing. You'll receive a new assignment shortly — we're redirecting to bold Round 48 experiments that work on the pre-ANP baseline, plus one student will re-implement the ANP from the W&B config and experiment description..
